### DISCUSSION TITLE
Welcome to photokit Discussions!

### DISCUSSION THREAD

**Discussion by RhetTbull** [RhetTbull/photokit#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by eblume** [comment]
Hi @RhetTbull ! I've got a project that handles some of my personal automation. Long, long, story short, it looks like I'm going to scan the most recent photos in my iCloud photos and then do further processing on them if they match some criteria, as part of a workflow starting from capturing screenshots on iCloud-syncing devices.

I was all set to go learn PyObjC and the proper applescript dictionary but then I found your project! Fantastic! I haven't had a chance to use it yet, but I'm looking forward to it. I did a quick reading and I had one question:

**Does photokit currently support [NSPredicate on fetchAssets](https://developer.apple.com/documentation/photokit/phfetchoptions/1624709-predicate)?**

I'm not an Obj-C programmer but my reading of PyObjC and your code makes it seem like it should be possible to support it via, if nothing else, more arguments to [PhotoLibrary.assets](https://github.com/RhetTbull/photokit/blob/def9150bd8cc2965bde49ad940c7c7eeeadf4f04/photokit/photolibrary.py#L294). If this already exists at a different layer and I'm missing it, could you help point me there? And if not, would you be open to a PR?

**Reply by eblume** [reply]
I'm going to send in a followup in case you're feeling generous: how much work do you think it would be to build an asynchronous (or threaded) interface to [PHPhotoLibraryChangeObserver](https://developer.apple.com/documentation/photokit/phphotolibrarychangeobserver)? My current plan is to attempt to use that (or fallback to polling) and subclass [watchdog's BaseObserver](https://github.com/gorakhargosh/watchdog/blob/6cdf07efa04d2651a632e963fb464c2265c91b9c/src/watchdog/observers/api.py#L228) to provide a watchdog-compatible event dispatcher for new photos.

**Reply by RhetTbull** [reply]
Hi @eblume photokit does not yet support using a predicate though that would be a great feature -- I'll open an issue for this.

Regarding PHPhotoLibraryChangeObserver -- I'm sure this is possible though I've not looked into doing this so not sure how much work it would be. I've done very little async in either python or objective C so there would be a bit of learning curve. In another project, [textinator](https://github.com/RhetTbull/textinator), I create a Spotlight query and use [NotificationCenter](https://developer.apple.com/documentation/foundation/nsnotificationcenter) to handle notifications of updates. Something like that might work with PHPhotoLibraryChangeObserver.

https://github.com/RhetTbull/textinator/blob/22ab8371d38a142858f56bcc4f7d2b364244176a/src/textinator.py#L359-L378

Please note that photokit is still very much "alpha" code.  I've been busy on some features for another project, [osxphotos](https://github.com/RhetTbull/osxphotos) but am close to releasing the updates at which time I plan to turn my attention to photokit again as I've got some down time over the holidays. 

For extracting text, you could like at the [VisionKit](https://developer.apple.com/documentation/visionkit) [code I use in textinator](https://github.com/RhetTbull/textinator/blob/main/src/macvision.py). 

If your desired use case is to monitor for new photos and extract text from new images, you could periodically call osxphotos to do this.  osxphotos reads in the entire Photos database and this takes some time (could be 10s of seconds on an older Mac with a large library) so it's not something you'd run continuously but it might work to bootstrap your project while you're waiting for the PhotoKit work.

For example, save the code below as `new_text.py` and run it with `osxphotos run new_text.py`.  

Alternatively, you could run it directly from GitHub: 
`osxphotos run https://raw.githubusercontent.com/RhetTbull/osxphotos/main/examples/new_text.py` if you just want to try it out. 

```python
"""Query any photos since the last run and extract text from them."""

import datetime
import json

import osxphotos
from osxphotos.cli import echo, kvstore, query_command


@query_command()
def new_text(photos: list[osxphotos.PhotoInfo], **kwargs):
    """Query any photos since the last run and extract text from them."""
    kv = kvstore("osxphotos_new_text")
    results = []
    for photo in photos:
        if photo.uuid in kv:
            continue
        if text := photo.search_info.detected_text:
            data = {
                "photo": photo.original_filename,
                "path": photo.path,
                "uuid": photo.uuid,
                "date_added": photo.date_added.isoformat()
                if photo.date_added
                else None,
                "text": text,
            }
            results.append(data)
        kv[photo.uuid] = datetime.datetime.now().isoformat()
    kv.close()
    echo(json.dumps(results, indent=4))


if __name__ == "__main__":
    new_text()

```

This uses some helper commands from osxphotos to build a simple CLI that works with all the osxphotos query commands.  It stores the previously processed photos in a key-value store that will be located in `~/.local/share/osxphotos` and will output a list of JSON objects which could then be processed by another script.

```json
[
    {
        "photo": "IMG_5316.HEIC",
        "path": "/Users/user/Pictures/Photos Library.photoslibrary/originals/1/1F9B5010-526F-4320-BE60-091E84B0F559.heic",
        "uuid": "1F9B5010-526F-4320-BE60-091E84B0F559",
        "date_added": "2023-12-17T17:16:09.999613-08:00",
        "text": [
            "VICTORIA"
        ]
    },
    {
        "photo": "IMG_5331.HEIC",
        "path": "/Users/user/Pictures/Photos Library.photoslibrary/originals/3/3B55361F-24D7-47AD-AC00-CA74E8CD3B2B.heic",
        "uuid": "3B55361F-24D7-47AD-AC00-CA74E8CD3B2B",
        "date_added": "2023-12-21T13:15:13.629278-08:00",
        "text": [
            "California",
            "Enya"
        ]
    }
]
```

Running it with no options results in processing all photos. You could limit the date using one or more of the [osxphotos query options](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-query). For example, `osxphotos run new_text.py --added-in-last "7 days"`

**Reply by RhetTbull** [reply]
By the way, this uses the text detected by Photos and used when you search for text in the Photos search bar. If you want more control over the detected text, you could use [PhotoInfo.detected_text](https://rhettbull.github.io/osxphotos/API_README.html#a-name-photoinfo-detected-text-detected-text-confidence-threshold-text-detection-confidence-threshold-a) which allows you to specify a threshold for detection and will return strings of text (e.g. a whole sentence) instead of individual words.

You can play with this easily using `osxphotos repl` which launches a python repl with your Photos database loaded. Select an image with text in it and in the REPL, run:

```pycon
>>> selected[0].detected_text(confidence_threshold=0.3)
[('Use the link below to start', 1.0), ('Post your reply', 1.0)]

```

This is much slower as the text detection gets called at the time of execution vs just using the results Photos has already stored in the database but gives you more control.

**Reply by RhetTbull** [reply]
I just took a look at [watchdog](https://github.com/gorakhargosh/watchdog) which I wasn't familiar with. You could probably use this to monitor `~/Pictures/Photos Library.photoslibrary/database` for any changes then run osxphotos to get the new text.

**Reply by eblume** [reply]
I think you've nailed it. In my use case I'm looking specifically for new screenshots sent from an iOS device, so I'll want to filter on perhaps metadata or dimension or size. I'm actually planning on sending the images over to OpenAI's function-calling assistant API to extract tabular data from the Wallet app, so I don't really need local OCR, although I will definitely check to see if detected_text is good enough for my use, thanks for the tip!

I appreciate the thorough response and in kind, here's the rambling story of what I'm up to. Uninterested readers can stop here.

In case you're curious, my plan is to re-use some watchdog-based code which was transcribing and then prompting an OpenAI assistant based on voice memos recorded via "Just Press Record", an iCloud-syncing voice memo program that works on multiple devices, giving me a very customizable and flexible [voice assistant](https://github.com/eblume/mole/blob/main/src/mole/voicememo.py#L56). Now I want to use watchdog and either photokit or perhaps osxphotos to build a similar workflow to export credit card transaction data from the iOS Wallet app by running a python CLI that prompts the user to start screenshotting their records, uploading them to a budgeting and financial tracking program (YNAB), intelligently zipping them together with existing records until the program recognizes it's 'caught up' to previously recorded transactions. All of this rigamarole seems to be necessary to get automated export from the wallet app PRIOR to the end of the month, which seems to be the earliest that the wallet app will let me export credit card info... other than literally screenshotting the info!

It's a bit of a Rube-Goldberg machine and while the existing functionality is already [open-sourced](https://github.com/eblume/mole) I am not expecting it to pan out to anything worth bragging about. But, it's [one of those things](https://xkcd.com/1205/) where if it saves me 30 minutes every week, it's worth 5 weeks (!!!) to fix it! Maybe I'll cook it up over the holiday and come back to you with my results.

**Reply by eblume** [reply]
Thanks again for the help! Cool library, I like all of your open source work!

**Reply by RhetTbull** [reply]
>  In my use case I'm looking specifically for new screenshots sent from an iOS device, so I'll want to filter on perhaps metadata or dimension or size. 

`osxphotos run new_text.py --screenshot` 

;-)

If you want to do this in your own code, you could do something like this:

```python
import osxphotos

options = osxphotos.QueryOptions(screenshot=True)
photos = osxphotos.PhotosDB().query(options)
```

**Reply by eblume** [reply]
Wow!! Perfect, thanks!

**Reply by RhetTbull** [reply]
@eblume OK, I have a *very rough* implementation of `registerChangeObserver:` implemented.  The following code for example, works to monitor for changes, returning a dataclass with added, removed, updated assets:

```python
"""Watch for changes in Photo Library"""

import time

import photokit


def change_observer(asset: photokit.AssetChanges):
    for a in asset.added:
        print(f"added: {a.uuid} {a.original_filename}")
    for a in asset.removed:
        print(f"removed: {a.uuid} {a.original_filename}")
    for a in asset.updated:
        print(f"updated: {a.uuid} {a.original_filename}")


def main():
    photolib = photokit.PhotoLibrary()
    photolib.observe_changes(change_observer)
    while True:
        # don't use sleep in a real app
        time.sleep(1)

if __name__ == "__main__":
    main()

```

output:
```
removed: 412CD294-3AFA-4FBF-8DCD-DFA427FCBBAB None
updated: 412CD294-3AFA-4FBF-8DCD-DFA427FCBBAB None
updated: 02914316-89F5-4ABF-A111-2B256113017A IMG_5366.HEIC
updated: 02914316-89F5-4ABF-A111-2B256113017A IMG_5366.HEIC
updated: 02914316-89F5-4ABF-A111-2B256113017A IMG_5366.HEIC
added: 0FDBEC82-5632-411D-BC43-B6B96E3E2CC6 Screenshot 2023-11-20 at 8.18.50 PM.png
removed: 412CD294-3AFA-4FBF-8DCD-DFA427FCBBAB None
```

I haven't figured out how to filter the duplicate notifications yet. Once I add support for predicates (#2) this will monitor only for changes to your search predicates. The predicates work will take some time because under the hood, `photokit` is actually two different libraries and I try to make it transparent to the user. One instance uses the public PhotoKit API which only works with the system photo library. The other uses undocumented private "multi library mode" which works with any library. But this has taken reverse engineering and implementation is slow and the method calls are all different than the public API. I want to find a way to make predicates as pythonic as possible and also work in both modes.

**Reply by eblume** [reply]
That looks good! For the observer, I'm curious, how is photokit handling whatever sort of objective-c/Grand Central Dispatch sort of work is necessary for the callback to fire? Is it tied to the python process itself somehow? If I'm troubleshooting and changes aren't coming through the bridge, do I just restart python or is there some thread I can kill and restart, etc.? And most importantly - where is this documented? Is it related somehow to `dispatch` ([pyobjc](https://pyobjc.readthedocs.io/en/latest/apinotes/libdispatch.html)/[apple](https://developer.apple.com/documentation/dispatch?language=objc))? Much appreciated, the more I seek to automate my mac the more I'm realizing that bringing objc and applescript in to my python source is an... *effective* idea. Let's go with "effective", I don't know if "good" is there yet for me. :)

**Reply by RhetTbull** [reply]
I think it's a good idea :-) See my other work, [macnotesapp](https://github.com/RhetTbull/macnotesapp) which wraps the Notes.app AppleScript interface and [photoscript](https://github.com/RhetTbull/PhotoScript) which does the same for Photos.

This isn't in the code on Github yet...still in a dev branch I'm working on. I need to get the API figured out, figure out how to observe the right fetch results, etc. As for how this works...at the moment I don't know exactly how and how this works with multiple threads etc. Currently, the PhotoLibrary class implements the PHPhotoLibraryChangeObserver  protocol then calls the registered callback when it receives a notifciation via the Objective C bridge. In the current implementation, it "works on my machine" in a single thread. Lots more work to get this ready for release.

In the meantime, if you are just observing for screenshots, you could do something like this:

```python
    photolib = photokit.PhotoLibrary()
    screenshots = photolib.smart_album(album_type=photokit.PhotoLibrarySmartAlbumType.Screenshots)
```

Then keep a log of every UUID in the screenshots collection (which will be a PhotoKit.Album class) and periodically check for new UUIDs. It's not elegant but might work well enough and maybe better than calling out to osxphotos.

**Reply by RhetTbull** [reply]
> the more I seek to automate my mac the more I'm realizing that bringing objc and applescript in to my python source is an... effective idea.

In addition to my projects referenced above, you might also take a look at [PyXA](https://github.com/SKaplanOfficial/PyXA) which wraps a number of Mac frameworks in Python as well as [Hammerspoon](https://www.hammerspoon.org/) which does the same with a Lua interface. I've not used Hammerspoon but have adopted some of the ideas in PyXA in my own projects (though I find the API non-intuitive). I have a design in my head for a "mac automation toolkit" written in Python but alas only so many hours to work on open source so that will remain just an idea for now. I have found that learning just enough objective C to be able to effectively use pyobjc from python is a very powerful tool for making things easier on the Mac.
